types/errors/errors.go

```diff
@@ -263,6 +263,16 @@ func (e *wrappedError) Cause() error {
 	return e.parent
 }
 
+// Is implements the built-in errors.Is
+func (e *wrappedError) Is(target error) bool {
```

Comment:
Yes, I see that I misunderstood you in the issue. Can you help a bit here what do you mean under the "work up the entire error chain"? 
As I get the Wrap get's one error and a description as string. The Error method compose them as a new error but it's just a simple string concatenation. So I don't get how can I getting back the chain. 
If the Wrap accepts two error instead of error plus string, so it can maintain a chain. On the principle side add this error to this error chain. 

So can you please clarify that please?

Reply:
What I mean is if I do the following:

```go
sdkerrors.Wrap(Err3, sdkerrors.Wrap(Err2, sdkerrors.Wrap(Err1, "...")))
```

If I discover an error `err` from the above, the following should be true

```
sdkerrors.Is(err, Err1) // => true
sdkerrors.Is(err, Err2) // => true
sdkerrors.Is(err, Err3) // => true
sdkerrors.Is(err, ErrSomeOther) // => false
```

See Go's documentation on `Is`: https://golang.org/pkg/errors/#Is

Apologies if this is already covered by your code, I'm just checking. We should have a testcase for this.

Reply:
It's nearly impossible to do that on the current implementation (on a clean code way), because it doesn't maintain a chain. 
Go's documentation says:
> The chain consists of err itself followed by the sequence of errors obtained by repeatedly calling Unwrap.

But we have an error and a string. So following your example it looks like the following: 
```
sdkerrors.Wrap(Err3, sdkerrors.Wrap(Err2, sdkerrors.Wrap(Err1, "...").Error()).Error())
```

Let me redesign it and based on your feedback we will discuss whether keep the changes or find an other way.

Reply:
Update about the changes. 
- I changed the Wrap parameter list from (error, string) to (error, error). In that case it can maintain a chain of errors. 
- Changed all of the occurrence of it.
- Separate the stack error and the original error to two different field just to keep the type and descriptor of the original error.

Suggestions:
- I want to remove the `Wrapf` function because inconsistency reasons and let the other side to use `fmt.Errorf` for that purpose.
- Add an `sdkerrors.New()` with the same functionality as the built-in errors.New because currently I should have import the `errors` package in every case. 

@alexanderbez Please let me know your thoughts.

Reply:
> It's nearly impossible to do that on the current implementation (on a clean code way), because it doesn't maintain a chain.

It _does_ maintain a chain. Sorry, maybe I wasn't clear or my example was confusing.

This is how `Is` should be implemented:

```go
// Is reports whether any error in e's chain matches a target.
func (e *wrappedError) Is(target error) bool {
	if target == nil {
		return e == target
	}

	w := e.Cause()

	for {
		if w == target {
			return true
		}

		x, ok := w.(causer)
		if ok {
			w = x.Cause()
		}
		if x == nil {
			return false
		}
	}
}
```

Tests:

```go
func TestWrappedIs(t *testing.T) {
	err := Wrap(ErrTxTooLarge, "context")
	require.True(t, stdlib.Is(err, ErrTxTooLarge))

	err = Wrap(err, "more context")
	require.True(t, stdlib.Is(err, ErrTxTooLarge))

	err = Wrap(err, "even more context")
	require.True(t, stdlib.Is(err, ErrTxTooLarge))
}
```

Please undo other changes (for now).

Reply:
Probably we misunderstood each other. I though under chain that a chain of the errors. It's actually creating the error message as a chain, but it can't handle the chain of the errors. Or I don't see that. 

My thoughts was the following: [link to goplayground](https://play.golang.org/p/YdsGd170Gb-)
Sorry about that.

Reply:
Well the messages have nothing to do with the errors -- essentially you're checking against the chain of `error` until matched or you get to `nil` 👍